File: Two_Layer_Triad_Resonance/dispersion_resonance_gui.py
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.widgets import Slider, TextBox

# =============================================================================
# 1. USER CONFIGURATION (EDIT HERE)
# =============================================================================
CONFIG = {
    # --- Initial Simulation Values ---
    'init': {
        'k0': 1.5,      # Shift Point (kappa_0)
        'R':  0.85,     # Density Ratio (rho1/rho2)
        'h':  1.0       # Depth Ratio (h1/h2)
    },
    
    # --- Initial Plot Limits ---
    'limits': {
        'xlim': 8.0,    # Plot Range X (+/- kappa)
        'ymax': 4.0,    # Y Max (Omega)
        'ymin': 0.0     # Y Min (Omega)
    },

    # --- Slider Ranges [Min, Max, Step] ---
    'ranges': {
        'k0':   [0.1,   5.0,   0.01],
        'R':    [0.01,  0.99,  0.001],
        'h':    [0.1,   5.0,   0.01],
        'xlim': [2.0,   20.0,  0.1],
        'ymax': [0.1,   15.0,  0.1],
        'ymin': [-15.0, 5.0,   0.1]
    },

    # --- Visual Styles ---
    'style': {
        'orig_surf':  {'color': 'b', 'linestyle': '-',  'linewidth': 3},
        'orig_int':   {'color': 'r', 'linestyle': '--', 'linewidth': 3},
        'shift_surf': {'color': 'b', 'linestyle': '-',  'linewidth': 1.0},
        'shift_int':  {'color': 'r', 'linestyle': ':',  'linewidth': 1.5}
    }
}

# =============================================================================
# 2. MATH KERNEL
# =============================================================================
def solve_dispersion(k_in, R, h):
    k = np.array(k_in)
    k[np.abs(k) < 1e-9] = 1e-9
    
    th_k = np.tanh(k)
    th_kh = np.tanh(k * h)
    coth_k = 1.0 / th_k
    coth_kh = 1.0 / th_kh
    
    A = R + (coth_k * coth_kh)
    B = -k * (coth_k + coth_kh)
    C = (k**2) * (1 - R)
    
    discriminant = np.sqrt(B**2 - 4 * A * C)
    
    Omega2_1 = (-B + discriminant) / (2 * A)
    Omega2_2 = (-B - discriminant) / (2 * A)
    
    Omega2_1[Omega2_1 < 0] = 0
    Omega2_2[Omega2_2 < 0] = 0
    
    Om_Surf = np.sqrt(Omega2_1)
    Om_Int  = np.sqrt(Omega2_2)
    
    # Fix k=0
    mask_zero = np.abs(k_in) < 1e-8
    Om_Surf[mask_zero] = 0
    Om_Int[mask_zero]  = 0
    
    return Om_Surf, Om_Int

def find_intersections(x_vec, y1_vec, y2_vec):
    diff = y1_vec - y2_vec
    sign_changes = np.where(np.diff(np.sign(diff)))[0]
    
    x_int = []
    y_int = []
    
    for idx in sign_changes:
        x_a, x_b = x_vec[idx], x_vec[idx+1]
        d_a, d_b = diff[idx], diff[idx+1]
        
        if d_a == d_b: continue 
            
        frac = abs(d_a) / (abs(d_a) + abs(d_b))
        x_val = x_a + frac * (x_b - x_a)
        
        val_a, val_b = y1_vec[idx], y1_vec[idx+1]
        y_val = val_a + frac * (val_b - val_a)
        
        x_int.append(x_val)
        y_int.append(y_val)
        
    return x_int, y_int

# =============================================================================
# 3. GUI SETUP
# =============================================================================
class ControlRow:
    def __init__(self, fig, y_pos, label_latex, key, update_func):
        # Unpack config for this specific key
        val_min, val_max, step = CONFIG['ranges'][key]
        
        # Determine initial value based on key type
        if key in CONFIG['init']:
            val_init = CONFIG['init'][key]
        else:
            val_init = CONFIG['limits'][key]

        self.step = step
        self.val_min = val_min
        self.val_max = val_max
        self.update_func = update_func

        # Label
        fig.text(0.02, y_pos + 0.055, label_latex, fontsize=11, fontweight='bold', ha='left')

        # Slider
        ax_slider = plt.axes([0.02, y_pos, 0.20, 0.03])
        # We allow default formatting initially to avoid TypeError
        self.slider = Slider(ax_slider, '', val_min, val_max, valinit=val_init, valstep=step)
        
        # --- FIX: Hide the number outside the slider manually ---
        self.slider.valtext.set_visible(False)
        
        # TextBox
        ax_box = plt.axes([0.23, y_pos, 0.08, 0.04])
        self.textbox = TextBox(ax_box, '', initial=str(val_init))

        self.slider.on_changed(self.on_slider_change)
        self.textbox.on_submit(self.on_text_submit)

    def on_slider_change(self, val):
        self.textbox.eventson = False 
        self.textbox.set_val(f"{val:.3f}")
        self.textbox.eventson = True
        self.update_func(None)

    def on_text_submit(self, text):
        try:
            val = float(text)
            if val < self.val_min: val = self.val_min
            if val > self.val_max: val = self.val_max
            
            self.slider.eventson = False
            self.slider.set_val(val)
            self.slider.eventson = True
            self.update_func(None)
        except ValueError:
            self.textbox.set_val(str(self.slider.val))

    @property
    def val(self):
        return self.slider.val

# Create Figure
fig = plt.figure(figsize=(14, 9))
plt.subplots_adjust(left=0.40, bottom=0.10, right=0.95, top=0.95)

# Axis Setup
ax = fig.add_subplot(111)
ax.set_xlabel(r'$\kappa$', fontsize=20)
ax.set_ylabel(r'$\Omega$', fontsize=20)
ax.tick_params(axis='both', which='major', labelsize=14)
ax.grid(True)

# Prepare Line Objects using Config Styles
lines = {}
lines['orig_s1'], = ax.plot([], [], **CONFIG['style']['orig_surf'])
lines['orig_s2'], = ax.plot([], [], **CONFIG['style']['orig_surf'])
lines['orig_i1'], = ax.plot([], [], **CONFIG['style']['orig_int'])
lines['orig_i2'], = ax.plot([], [], **CONFIG['style']['orig_int'])

lines['shift_s1'], = ax.plot([], [], **CONFIG['style']['shift_surf'])
lines['shift_s2'], = ax.plot([], [], **CONFIG['style']['shift_surf'])
lines['shift_i1'], = ax.plot([], [], **CONFIG['style']['shift_int'])
lines['shift_i2'], = ax.plot([], [], **CONFIG['style']['shift_int'])

# --- MODIFIED MARKERS ---
# 1. Origin Point (Fixed Black Circle)
pt_origin, = ax.plot([0], [0], 'ko', markersize=11, zorder=11)

# 2. Shift Center Point (Black Circle)
pt_center, = ax.plot([], [], 'ko', markersize=11, zorder=11)

# 3. Other Intersections (Yellow Circles, NO black edge)
# Changed markeredgecolor='k' to markeredgecolor='none'
# Use 'o' for circle, pure Hex Yellow '#FFFF00', and size 12 (30% bigger)
pt_intersect, = ax.plot([], [], 'o', color='#FFFF00', markersize=12, markeredgecolor='k', zorder=10)

# Legend Info
info_text = (
    "Thick Blue: Orig. Surface\n"
    "Thick Red:  Orig. Internal\n"
    "Thin Blue:  Shifted Surface\n"
    "Thin Red:   Shifted Internal\n"
    "Black O:    Origin & Shift Point\n"
    "Yellow O:   Other Intersections"
)
fig.text(0.02, 0.15, info_text, fontsize=10, style='italic', va='top', bbox=dict(facecolor='white', alpha=0.5))
fig.text(0.02, 0.96, "Control Panel", fontsize=14, fontweight='bold')

# =============================================================================
# 4. UPDATE LOGIC
# =============================================================================
def update(val):
    xl = ctrl_xlim.val
    yl_max = ctrl_ymax.val
    yl_min = ctrl_ymin.val
    k0 = ctrl_k0.val
    R = ctrl_R.val
    h = ctrl_h.val

    if yl_min >= yl_max: yl_min = yl_max - 0.1

    calc_limit = xl * 1.1
    k_vec = np.linspace(-calc_limit, calc_limit, 1000)

    # 1. Original
    Om_S_orig, Om_I_orig = solve_dispersion(k_vec, R, h)
    Orig_Y = [Om_S_orig, -Om_S_orig, Om_I_orig, -Om_I_orig]

    # 2. Shift Point
    Om_S_0, _ = solve_dispersion([k0], R, h)
    Omega0 = Om_S_0[0]

    # 3. Shifted
    k_shift = k_vec - k0
    Om_S_shift, Om_I_shift = solve_dispersion(k_shift, R, h)
    Shift_Y = [
        Omega0 + Om_S_shift, Omega0 - Om_S_shift,
        Omega0 + Om_I_shift, Omega0 - Om_I_shift
    ]

    # Update Lines
    lines['orig_s1'].set_data(k_vec, Orig_Y[0])
    lines['orig_s2'].set_data(k_vec, Orig_Y[1])
    lines['orig_i1'].set_data(k_vec, Orig_Y[2])
    lines['orig_i2'].set_data(k_vec, Orig_Y[3])

    lines['shift_s1'].set_data(k_vec, Shift_Y[0])
    lines['shift_s2'].set_data(k_vec, Shift_Y[1])
    lines['shift_i1'].set_data(k_vec, Shift_Y[2])
    lines['shift_i2'].set_data(k_vec, Shift_Y[3])

    # Update Shift Center Point (Black)
    pt_center.set_data([k0], [Omega0])

    # Find Intersections
    ix_vals = []
    iy_vals = []

    for y1 in Orig_Y:
        for y2 in Shift_Y:
            tx, ty = find_intersections(k_vec, y1, y2)
            for x_i, y_i in zip(tx, ty):
                if yl_min <= y_i <= yl_max:
                    # Filter out the origin and the shift point from the yellow list
                    tol = 1e-5
                    is_origin = (abs(x_i) < tol and abs(y_i) < tol)
                    is_center = (abs(x_i - k0) < tol and abs(y_i - Omega0) < tol)
                    
                    if not is_origin and not is_center:
                        ix_vals.append(x_i)
                        iy_vals.append(y_i)

    # Update Yellow Intersections
    pt_intersect.set_data(ix_vals, iy_vals)
    
    ax.set_xlim(-xl, xl)
    ax.set_ylim(yl_min, yl_max)
    fig.canvas.draw_idle()

# =============================================================================
# 5. INITIALIZE CONTROLS
# =============================================================================
curr_y = 0.85
def next_y():
    global curr_y
    y = curr_y
    curr_y -= 0.10
    return y

# Generate Controls (Order matters for layout)
ctrl_xlim = ControlRow(fig, next_y(), r'Plot Range $X (\pm\kappa)$', 'xlim', update)
ctrl_ymax = ControlRow(fig, next_y(), r'Y Max ($\Omega$)', 'ymax', update)
ctrl_ymin = ControlRow(fig, next_y(), r'Y Min ($\Omega$)', 'ymin', update)

# Divider
line_ax = plt.axes([0.02, curr_y + 0.05, 0.30, 0.002])
line_ax.set_xticks([]); line_ax.set_yticks([]); line_ax.set_facecolor('black')

ctrl_k0 = ControlRow(fig, next_y(), r'Shift Point $(\kappa_0)$', 'k0', update)
ctrl_R  = ControlRow(fig, next_y(), r'Density Ratio $(R)$', 'R', update)
ctrl_h  = ControlRow(fig, next_y(), r'Depth Ratio $(h)$', 'h', update)

update(None)
plt.show()

File: Two_Layer_Triad_Resonance/dispersion_resonance_gui.m
function dispersion_resonance_gui
    % Two-Layer Fluid Dispersion: Resonance Analysis & GUI
    % 
    
    clear; clc; close all;
    % =========================================================================
    % 1. CONTROL PANEL (CONFIGURATION)
    % =========================================================================
    
    % --- Initial Values ---
    cfg.k0_init     = 1.5;    % Reference Wavenumber
    cfg.R_init      = 0.85;   % Density Ratio
    cfg.h_init      = 1.0;    % Depth Ratio
    
    cfg.xlim_init   = 8.0;    % X-Axis Limit (+/-)
    cfg.ymax_init   = 4.0;    % Y-Axis Upper Limit
    cfg.ymin_init   = 0.0;   % Y-Axis Lower Limit
    
    cfg.num_points  = 1000;   % Grid resolution
    
    % --- Slider Settings: [Min, Max, Step_Size] ---
    cfg.k0_sets     = [0.1,   5.0,   0.01];   
    cfg.R_sets      = [0.01,  0.99,  0.001]; 
    cfg.h_sets      = [0.1,   5.0,   0.01];  
    
    cfg.xlim_sets   = [2.0,   20.0,  0.1];   
    cfg.ymax_sets   = [0.1,   15.0,  0.1];    
    cfg.ymin_sets   = [-15.0, 0.0,   0.1];    
    
    % --- Visual Styles ---
    cfg.style_orig_surf  = {'Color', 'b', 'LineWidth', 3, 'LineStyle', '-'};
    cfg.style_orig_int   = {'Color', 'r', 'LineWidth', 3, 'LineStyle', '--'};
    cfg.style_shift_surf = {'Color', 'b', 'LineWidth', 0.5, 'LineStyle', '-'};
    cfg.style_shift_int  = {'Color', 'r', 'LineWidth', 1.5, 'LineStyle', ':'}; 
    
    % =========================================================================
    % 2. GUI SETUP
    % =========================================================================
    
    fig = figure('Name', 'Resonance Analysis v7', ...
                 'Color', 'w', 'Units', 'pixels', 'Position', [100, 100, 1200, 800]);
             
    % Control Panel Container
    controlPanel = uipanel('Parent', fig, 'Title', 'Control Panel', ...
                           'FontSize', 12, 'BackgroundColor', 'w', ...
                           'Position', [0.02 0.05 0.25 0.9]);
    
    % --- THE TRICK: Invisible Axes for LaTeX Text ---
    % Standard uicontrols don't support LaTeX. We create an invisible axes
    % spanning the control panel [0,1]x[0,1] to draw text() objects on it.
    lbl_ax = axes('Parent', controlPanel, 'Position', [0 0 1 1], ...
                  'Visible', 'off', 'XLim', [0 1], 'YLim', [0 1]);
                       
    % --- Helper to create Slider + Edit Box + LaTeX Label ---
    function [sld, edt] = create_control_row(parent, label_axes, y_pos, latex_str, settings, init_val, plot_update_func)
        % settings = [min, max, step]
        min_v = settings(1); max_v = settings(2); step_v = settings(3);
        
        % Normalize step
        range_v = max_v - min_v;
        if range_v <= 0, range_v = 1; end
        norm_step = [step_v/range_v, (step_v*10)/range_v];
        
        % 1. LaTeX Label (Using text() on the invisible axes)
        % y_pos is normalized (0 to 1). We offset slightly up for the label.
        % UPDATED: FontSize changed to 12
        text(label_axes, 0.05, y_pos + 0.055, latex_str, ...
             'Interpreter', 'latex', 'FontSize', 12, 'FontWeight', 'bold', ...
             'Color', 'k', 'HorizontalAlignment', 'left');
              
        % 2. Edit Box
        edt = uicontrol(parent, 'Style', 'edit', 'String', num2str(init_val), ...
                  'Units', 'normalized', 'Position', [0.70, y_pos+0.032, 0.25, 0.04], ...
                  'BackgroundColor', 'w', 'ForegroundColor', 'k', ...
                  'Callback', {@edit_callback});
              
        % 3. Slider
        sld = uicontrol(parent, 'Style', 'slider', ...
                  'Min', min_v, 'Max', max_v, 'Value', init_val, ...
                  'SliderStep', norm_step, ...
                  'UserData', struct('step', step_v, 'editBox', edt), ... 
                  'Units', 'normalized', 'Position', [0.05, y_pos, 0.9, 0.03], ...
                  'Callback', {@slider_callback});
              
        % --- Local Callbacks ---
        function slider_callback(src, ~)
            data = get(src, 'UserData');
            val = get(src, 'Value');
            val = round(val / data.step) * data.step; 
            set(src, 'Value', val);
            set(data.editBox, 'String', num2str(val));
            plot_update_func();
        end
        function edit_callback(src, ~)
            str = get(src, 'String');
            val = str2double(str);
            if isnan(val), val = get(sld, 'Value'); 
            else
                if val < min_v, val = min_v; end
                if val > max_v, val = max_v; end
            end
            set(sld, 'Value', val);
            set(src, 'String', num2str(val));
            plot_update_func();
        end
    end

    % --- Create Controls ---
    % Now we pass 'lbl_ax' and write full LaTeX strings
    
    [sld_xlim, edt_xlim] = create_control_row(controlPanel, lbl_ax, 0.92, ...
        'Plot Range $X (\pm\kappa)$', cfg.xlim_sets, cfg.xlim_init, @updatePlot);
    
    [sld_ymax, edt_ymax] = create_control_row(controlPanel, lbl_ax, 0.84, ...
        'Y Max ($\Omega$)', cfg.ymax_sets, cfg.ymax_init, @updatePlot);
        
    [sld_ymin, edt_ymin] = create_control_row(controlPanel, lbl_ax, 0.76, ...
        'Y Min ($\Omega$)', cfg.ymin_sets, cfg.ymin_init, @updatePlot);
    
    % Divider
    uicontrol(controlPanel, 'Style', 'frame', 'Units', 'normalized', ...
              'Position', [0.02 0.73 0.96 0.002], 'BackgroundColor', [0.8 0.8 0.8]); 
    
    [sld_k0, edt_k0] = create_control_row(controlPanel, lbl_ax, 0.65, ...
        'Shift Point $(\kappa_0)$', cfg.k0_sets, cfg.k0_init, @updatePlot);
        
    [sld_R,  edt_R]  = create_control_row(controlPanel, lbl_ax, 0.55, ...
        'Density Ratio $(R = \rho_1/\rho_2)$', cfg.R_sets, cfg.R_init, @updatePlot);
        
    [sld_h,  edt_h]  = create_control_row(controlPanel, lbl_ax, 0.45, ...
        'Depth Ratio $(h = h_1/h_2)$', cfg.h_sets, cfg.h_init, @updatePlot);
    
    % Legend Info
    infoStr = sprintf([...
        'Thick Blue: Orig. Surface\n' ...
        'Thick Red:  Orig. Internal\n' ...
        'Thin Blue:  Shifted Surface\n' ...
        'Thin Yell:  Shifted Internal\n' ...
        'Green O:    All Intersections']);
    uicontrol(controlPanel, 'Style', 'text', 'String', infoStr, ...
              'Units', 'normalized', 'Position', [0.05 0.1 0.9 0.25], ...
              'BackgroundColor', 'w', 'HorizontalAlignment', 'left', ...
              'FontAngle', 'italic', 'FontSize', 10);
          
    % =========================================================================
    % 3. PLOTTING AREA
    % =========================================================================
    
    ax = axes('Parent', fig, 'Position', [0.35 0.15 0.6 0.8]);
    hold(ax, 'on'); grid(ax, 'on'); box(ax, 'on');
    
    xlabel(ax, '$\kappa$', 'Interpreter', 'latex', 'FontSize', 20);
    ylabel(ax, '$\Omega$', 'Interpreter', 'latex', 'FontSize', 20);
    set(ax, 'FontSize', 20, 'TickLabelInterpreter', 'latex');
    
    updatePlot();
    
    % =========================================================================
    % 4. UPDATE LOGIC
    % =========================================================================
    function updatePlot(~, ~)
        xl_val   = get(sld_xlim, 'Value');
        ymax_val = get(sld_ymax, 'Value');
        ymin_val = get(sld_ymin, 'Value');
        k0_val   = get(sld_k0, 'Value');
        R_val    = get(sld_R, 'Value');
        h_val    = get(sld_h, 'Value');
        
        if ymin_val >= ymax_val, ymin_val = ymax_val - 0.1; end
        
        % Calculation
        calc_limit = xl_val * 1.1; 
        k_vec = linspace(-calc_limit, calc_limit, cfg.num_points);
        
        [Om_S_orig, Om_I_orig] = solve_dispersion(k_vec, R_val, h_val);
        Orig_Y = {Om_S_orig, -Om_S_orig, Om_I_orig, -Om_I_orig};
        
        [Om_S_0, ~] = solve_dispersion(k0_val, R_val, h_val);
        Omega0 = Om_S_0;
        
        k_shift = k_vec - k0_val;
        [Om_S_shift, Om_I_shift] = solve_dispersion(k_shift, R_val, h_val);
        Shift_Y = { ...
            Omega0 + Om_S_shift, Omega0 - Om_S_shift, ...
            Omega0 + Om_I_shift, Omega0 - Om_I_shift ...
        };
        
        % Plotting
        cla(ax);
        
        % Orig
        plot(ax, k_vec, Orig_Y{1}, cfg.style_orig_surf{:});
        plot(ax, k_vec, Orig_Y{2}, cfg.style_orig_surf{:});
        plot(ax, k_vec, Orig_Y{3}, cfg.style_orig_int{:});
        plot(ax, k_vec, Orig_Y{4}, cfg.style_orig_int{:});
        
        % Shifted
        for i = 1:4
            if i <= 2
                plot(ax, k_vec, Shift_Y{i}, cfg.style_shift_surf{:});
            else
                plot(ax, k_vec, Shift_Y{i}, cfg.style_shift_int{:});
            end
        end
        
        % Point P
        plot(ax, k0_val, Omega0, 'ko', 'MarkerFaceColor', 'k', 'MarkerSize', 8);
        
        % Intersections
        for i = 1:4
            y1 = Orig_Y{i};
            for j = 1:4
                y2 = Shift_Y{j};
                diff_v = y1 - y2;
                sign_chg = (diff_v(1:end-1) .* diff_v(2:end)) <= 0;
                idx_list = find(sign_chg);
                
                for k_idx = 1:length(idx_list)
                    id = idx_list(k_idx);
                    x_a = k_vec(id); x_b = k_vec(id+1);
                    d_a = diff_v(id); d_b = diff_v(id+1);
                    frac = abs(d_a)/(abs(d_a)+abs(d_b));
                    k_int = x_a + frac*(x_b - x_a);
                    
                    val_a = y1(id); val_b = y1(id+1);
                    om_int = val_a + frac*(val_b - val_a);
                    
                    if om_int >= ymin_val && om_int <= ymax_val
                         plot(ax, k_int, om_int, 'go', 'LineWidth', 2, 'MarkerSize', 9);
                    end
                end
            end
        end
        
        xlim(ax, [-xl_val, xl_val]);
        ylim(ax, [ymin_val, ymax_val]);
    end
    
    function [Om_Surf, Om_Int] = solve_dispersion(k_in, R, h)
        k = k_in; 
        k(abs(k) < 1e-9) = 1e-9;
        
        th_k  = tanh(k);
        th_kh = tanh(k * h);
        coth_k  = 1 ./ th_k;
        coth_kh = 1 ./ th_kh;
        
        A = R + (coth_k .* coth_kh);
        B = -k .* (coth_k + coth_kh);
        C = (k.^2) * (1 - R);
        
        discriminant = sqrt(B.^2 - 4 .* A .* C);
        
        Omega2_1 = (-B + discriminant) ./ (2 * A);
        Omega2_2 = (-B - discriminant) ./ (2 * A);
        
        Omega2_1(Omega2_1 < 0) = 0;
        Omega2_2(Omega2_2 < 0) = 0;
        
        Om_Surf = sqrt(Omega2_1);
        Om_Int  = sqrt(Omega2_2);
        
        Om_Surf(abs(k_in) < 1e-8) = 0;
        Om_Int(abs(k_in) < 1e-8)  = 0;
    end
end
